src/object/srv_obj.c

```diff
@@ -791,58 +734,235 @@ ds_obj_check_dtx_config(uint32_t flags, struct dtx_id *dti)
+	if ((flags != 0 || dti != NULL) && obj_modify_rpc(opc) &&
+	     !daos_oc_echo_type(daos_obj_id2class(oid.id_pub))) {
+		rc = ds_obj_check_dtx_config(flags, dti);
+		if (rc != 0)
+			D_GOTO(out_put, rc);
+	}
+
+out_put:
+	if (rc) {
+		if (*contp != NULL) {
+			ds_cont_child_put(*contp);
+			*contp = NULL;
+		}
+		if (*hdlp != NULL) {
+			ds_cont_hdl_put(*hdlp);
+			*hdlp = NULL;
+		}
+	}
+
+	return rc;
+}
+
+void
+ds_obj_tgt_update_handler(crt_rpc_t *rpc)
+{
+	struct obj_rw_in		*orw = crt_req_get(rpc);
+	struct obj_rw_out		*orwo = crt_reply_get(rpc);
+	struct ds_cont_hdl		*cont_hdl = NULL;
+	struct ds_cont_child		*cont = NULL;
+	struct dtx_handle		*dth = NULL;
+	struct dtx_conflict_entry	 conflict = { 0 };
+	uint32_t			 map_ver = 0;
+	int				 rc;
+
+	D_ASSERT(orw != NULL);
+	D_ASSERT(orwo != NULL);
+
+	rc = ds_pre_check(orw->orw_oid, orw->orw_map_ver, orw->orw_co_hdl,
+			  orw->orw_co_uuid, opc_get(rpc->cr_opc),
+			  orw->orw_flags, &orw->orw_dti,
+			  &cont_hdl, &cont);
+	if (rc)
+		goto out;
+
+	D_ASSERT(cont_hdl->sch_pool != NULL);
+	map_ver = cont_hdl->sch_pool->spc_map_version;
+
+	D_DEBUG(DB_TRACE, "rpc %p opc %d "DF_UOID" dkey %d %s tag/xs %d/%d eph "
+		DF_U64", pool ver %u/%u with "DF_DTI".\n",
+		rpc, opc_get(rpc->cr_opc), DP_UOID(orw->orw_oid),
+		(int)orw->orw_dkey.iov_len, (char *)orw->orw_dkey.iov_buf,
+		dss_get_module_info()->dmi_tgt_id,
+		dss_get_module_info()->dmi_xs_id, orw->orw_epoch,
+		orw->orw_map_ver, map_ver, DP_DTI(&orw->orw_dti));
+
+	/* Handle resend. */
+	if (orw->orw_flags & ORF_RESEND) {
+		rc = dtx_handle_resend(cont->sc_hdl, &orw->orw_oid,
+				       &orw->orw_dti,
+				       orw->orw_dkey_hash, false);
+		if (rc == -DER_ALREADY)
+			D_GOTO(out, rc = 0);
+	}
+
+	/**
+	 * XXX we should use different api for local dtx, will fix in the
+	 * following patch.
+	 **/
+	rc = dtx_begin(&orw->orw_dti, &orw->orw_oid, cont->sc_hdl,
+		       orw->orw_epoch, orw->orw_dkey_hash,
+		       &conflict, orw->orw_dti_cos.ca_arrays,
+		       orw->orw_dti_cos.ca_count, orw->orw_map_ver,
+		       DAOS_INTENT_UPDATE, false, &dth);
+	if (rc != 0) {
+		D_ERROR(DF_UOID": Failed to start DTX for update %d.\n",
+			DP_UOID(orw->orw_oid), rc);
+		D_GOTO(out, rc);
+	}
+
+	rc = obj_local_rw(rpc, cont_hdl, cont, dth);
+	if (rc != 0) {
+		D_ERROR(DF_UOID": rw_local (update) failed %d.\n",
+			DP_UOID(orw->orw_oid), rc);
+		D_GOTO(out, rc);
+	}
+
+	/**
+	 * XXX layer violation, it should be moved to DTX module, will
+	 * handle in following patches.
+	 */
+	if (orw->orw_dti_cos.ca_count > 0) {
+		/* XXX: For non-leader replica, even if we fail to
+		 *	make related modification for some reason,
+		 *	we still need to commit the DTXs for CoS.
+		 *	Because other replica may have already
+		 *	committed them. For leader case, it is
+		 *	not important even if we miss to commit
+		 *	the CoS DTXs, because they are still in
+		 *	CoS cache, and can be committed next time.
+		 */
+		rc = vos_dtx_commit(cont->sc_hdl,
```

Comment:
The DTXs in the @orw_dti_cos.ca_arrays may have been committed during the local update, so we only need this handling for local update failure case.

Reply:
yes, it is only in ds_obj_tgt_update_handler().